File: src/types/result-artifact.ts
export type ArtifactProcessingStatus =
  | 'raw'
  | 'parsing'
  | 'valid'
  | 'valid_with_warnings'
  | 'invalid';

export type ResultArtifactType =
  | 'generic_text'
  | 'generic_json'
  | 'chapter_text'
  | 'quality_report'
  | 'character_candidates'
  | 'event_candidates'
  | 'setting_candidates'
  | 'style_analysis'
  | 'chapter_summary'
  | 'volume_summary'
  | 'outline'
  | 'tool_result'
  | 'plan';

export interface ResultArtifact {
  artifactId: string;
  taskId: string;
  attemptId: string;
  sourceInputSnapshotId: string;
  artifactType: ResultArtifactType;
  schemaVersion: number;
  rawContentRefId: string;
  displayContentRefId?: string;
  displayContentHash?: string;
  structuredPayloadRefId?: string;
  structuredPayloadHash?: string;
  sourceNovelId: string;
  sourceChapterId?: string;
  sourceDraftId?: string;
  sourceDraftVersion?: number;
  sourceBaseContentHash?: string;
  contentHash: string;
  contentLength: number;
  processingStatus: ArtifactProcessingStatus;
  parentArtifactId?: string;
  derivationType?: string;
  createdAt: string;
}

export interface ArtifactValidationIssue {
  issueId: string;
  artifactId: string;
  validationRunId: string;
  issueIndex: number;
  severity: 'warning' | 'error';
  code: string;
  message: string;
  jsonPath?: string;
  detailsJson?: Record<string, unknown>;
  validatorVersion: string;
  createdAt: string;
}

export interface ResultArtifactBundle {
  artifact: ResultArtifact;
  rawContent: string;
  displayContent?: string;
  structuredPayloadJson?: unknown;
  issues: ArtifactValidationIssue[];
}

export interface CreateResultArtifactInput {
  taskId: string;
  attemptId: string;
  artifactType: ResultArtifactType;
  schemaVersion: number;
  rawContent: string;
  displayContent?: string;
  structuredPayloadJson?: unknown;
  parentArtifactId?: string;
  derivationType?: string;
}

File: src-tauri/src/domain/result_artifact.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ArtifactProcessingStatus {
    Raw,
    Parsing,
    Valid,
    ValidWithWarnings,
    Invalid,
}

impl ArtifactProcessingStatus {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Raw => "raw",
            Self::Parsing => "parsing",
            Self::Valid => "valid",
            Self::ValidWithWarnings => "valid_with_warnings",
            Self::Invalid => "invalid",
        }
    }
}

pub fn is_supported_artifact_type(value: &str) -> bool {
    matches!(
        value,
        "generic_text"
            | "generic_json"
            | "chapter_text"
            | "scene_text"
            | "quality_report"
            | "character_candidates"
            | "event_candidates"
            | "setting_candidates"
            | "style_analysis"
            | "chapter_summary"
            | "volume_summary"
            | "outline"
            | "tool_result"
            | "plan"
    )
}

pub fn is_supported_artifact_contract(value: &str, schema_version: i64) -> bool {
    schema_version == 1 && is_supported_artifact_type(value)
}

pub fn requires_json(value: &str) -> bool {
    matches!(
        value,
        "generic_json"
            | "quality_report"
            | "character_candidates"
            | "event_candidates"
            | "setting_candidates"
            | "style_analysis"
            | "chapter_summary"
            | "volume_summary"
            | "outline"
            | "tool_result"
            | "plan"
    )
}
